### PR TITLE
Feature 325

### LINKED ISSUE #325: As a guardian I would like to be able to copy weekplans so that I can easily reuse weekplans
**Describe the solution you'd like**
When the edit button is pressed a delete button is shown. A additional copy button should also pop up as shown in the picture below. Just as with the activities it should be possible to select more then one week. 
image

When the copy weekplan button is pressed a popup as shown below should popup.
image

If the "kopier til andre borgere" is pressed a screen with all the citizens as shown below should pop up. On this screen you can chose one or more citizens to copy the weeks too.  
image

When the weeks have been copied, the following pop up should appear.
image

If there already exist a weekplan with the same date as the one being copied in the destantion the following popup should appear. Where the ændr button should take you to the same screen as tilføj ugeplan with the information for the weekplan being copied.  
image

### PR DESCRIPTION
**Pull request by aerlingsson**
Fixes #325

### CODE REVIEW

**Review by mads-bo — COMMENTED** [review]
@aau-giraf/sw609f20 you have been assigned this. Here's your checklist.
Please use the following checklist. Feel free to check it off as you go along.

**Code Design**

- [x] The code is in the right place? (Both in terms of folder structure and class structure)
- [x] This code could not, to the best of my knowledge, have reused existing code
- [x] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems
- [x] The code does it introduce functionality that is not necessary for solving the problem

**Code Readability and Maintainability**

- [x] Names are meaningful and self-documenting
- [x] It is understandable, by reading the code, what it does
- [x] The code is simple and readable. i.e. it contains no "hack", or obscure solution
- [x] All new functionality is covered by tests
- [x] It is understandable, by reading the tests, what they do
- [ ] The tests covers sensible cases. Both happy paths and exception paths
- [ ] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing
- [ ] Error messages in the code are understandable
- [ ] Confusing sections of code are covered by comments

**Functionality**

For this part, open the app, and test:

- [ ] The features that the code claims to implement, are actually implemented
- [ ] The code does not look like it has bugs
- [ ] All the new screens are reachable through in-app navigation

**Points for Discussion and Reflection**

- [ ] UI messages, that the user might meet, are actually helpful
- [ ] There are no errors that are bad practice. E.g. hard-coding a test database connection
- [ ] There is well written code in this. Remember to tell the developer!

**Review by mads-bo — COMMENTED** [review]
@aau-giraf/sw613f20 you have been assigned this. Here's your checklist.
Please use the following checklist. Feel free to check it off as you go along.

**Code Design**

- [ ] The code is in the right place? (Both in terms of folder structure and class structure)
- [ ] This code could not, to the best of my knowledge, have reused existing code
- [ ] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems
- [ ] The code does it introduce functionality that is not necessary for solving the problem

**Code Readability and Maintainability**

- [ ] Names are meaningful and self-documenting
- [ ] It is understandable, by reading the code, what it does
- [ ] The code is simple and readable. i.e. it contains no "hack", or obscure solution
- [ ] All new functionality is covered by tests
- [ ] It is understandable, by reading the tests, what they do
- [ ] The tests covers sensible cases. Both happy paths and exception paths
- [ ] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing
- [ ] Error messages in the code are understandable
- [ ] Confusing sections of code are covered by comments

**Functionality**

For this part, open the app, and test:

- [ ] The features that the code claims to implement, are actually implemented
- [ ] The code does not look like it has bugs
- [ ] All the new screens are reachable through in-app navigation

**Points for Discussion and Reflection**

- [ ] UI messages, that the user might meet, are actually helpful
- [ ] There are no errors that are bad practice. E.g. hard-coding a test database connection
- [ ] There is well written code in this. Remember to tell the developer!

**Comment by aerlingsson** [comment]
Some checks fail after fixing merge conflicts. I'm looking into this

**Review by Lortemanden — DISMISSED** [review]

**Comment by nicolaibjerring** [comment]
The issue states that is should be possible to copy multiple weekplans at once, however this is not how it is implemented (we can see that you specifically test to make sure it is only possible to copy exactly one weekplan). Is this because you have an agreement with the PO group or?

**Comment by aerlingsson** [comment]
> 
> 
> The issue states that is should be possible to copy multiple weekplans at once, however this is not how it is implemented (we can see that you specifically test to make sure it is only possible to copy exactly one weekplan). Is this because you have an agreement with the PO group or?

Yes, PO specifically wanted it like this

**Review by Lortemanden — DISMISSED** [review]

**Review by Lortemanden — CHANGES_REQUESTED** [review]
If you try to copy to another weekplan to another citizen, but that citizen already has a weekplan for that week it will popup and ask if you want to change or overide it. If you chose to change and change the week or year the weekplan will not be copied to the citizen. Jeg har vedlagt en video hvor jeg viser fejlen.
[Kopier_Uge_bug.zip](https://github.com/aau-giraf/weekplanner/files/4545844/Kopier_Uge_bug.zip)

File: android/app/build.gradle [hunk 1]
```diff
@@ -38,7 +38,7 @@ android {
 
     defaultConfig {
         // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
-        applicationId "dk.aau.cs.giraf.weekplanner"
+        applicationId "com.example.weekplanner"
```
Comment (frederikspang) [inline android/app/build.gradle]: This is a pretty critical "Accidental" change.

File: .github/PULL_REQUEST_TEMPLATE/pull_request_template.md [hunk 2]
```diff
@@ -1,11 +0,0 @@
...
-name: Pull request template
-about: If you are making a pull request
-title: 'feature ISSUE_NUMBER'
-
----
-
-**Link issue the pull request fixes**
-fixes #Issue_Number
-**Link pull request to current giraf project**
-To the right, click link to project. Click on organizations and link to the current giraf project.
```
Comment (frederikspang) [inline .github/PULL_REQUEST_TEMPLATE/pull_request_template.md]: ?

File: ios/Flutter/Debug.xcconfig [hunk 3]
```diff
@@ -1,2 +1 @@
-#include "Pods/Target Support Files/Pods-Runner/Pods-Runner.debug.xcconfig"
```
Comment (frederikspang) [inline ios/Flutter/Debug.xcconfig:1]: ?

File: ios/Flutter/Release.xcconfig [hunk 4]
```diff
@@ -1,2 +1 @@
-#include "Pods/Target Support Files/Pods-Runner/Pods-Runner.release.xcconfig"
```
Comment (frederikspang) [inline ios/Flutter/Release.xcconfig]: ?

File: ios/Runner.xcworkspace/contents.xcworkspacedata [hunk 5]
```diff
@@ -4,7 +4,4 @@
    <FileRef
       location = "group:Runner.xcodeproj">
    </FileRef>
-   <FileRef
-      location = "group:Pods/Pods.xcodeproj">
-   </FileRef>
```
Comment (frederikspang) [inline ios/Runner.xcworkspace/contents.xcworkspacedata:9]: ?

File: lib/bootstrap.dart [hunk 6]
```diff
@@ -19,6 +19,8 @@ import 'package:api_client/api/api.dart';
 import 'package:weekplanner/providers/environment_provider.dart' as environment;
 import 'package:weekplanner/blocs/copy_activities_bloc.dart';
 
+import 'blocs/copy_resolve_bloc.dart';
```
Comment (frederikspang) [inline lib/bootstrap.dart]: Please follow `package:weekplanner` standard as above.

File: lib/screens/choose_citizen_screen.dart [hunk 7]
```diff
@@ -5,11 +5,12 @@ import 'package:weekplanner/blocs/choose_citizen_bloc.dart';
 import 'package:weekplanner/di.dart';
 import 'package:weekplanner/models/enums/app_bar_icons_enum.dart';
 import 'package:weekplanner/routes.dart';
-import 'package:weekplanner/screens/new_citizen_screen.dart';
 import 'package:weekplanner/screens/weekplan_selector_screen.dart';
 import 'package:weekplanner/widgets/citizen_avatar_widget.dart';
 import 'package:weekplanner/widgets/giraf_app_bar_widget.dart';
 
+import 'new_citizen_screen.dart';
```
Comment (frederikspang) [inline lib/screens/choose_citizen_screen.dart]: Please follow package:weekplanner standard as above.

File: lib/widgets/settings_widgets/settings_section_colorThemeButton.dart [hunk 8]
```diff
@@ -56,4 +56,4 @@ class SettingsColorThemeCheckMarkButton extends SettingsSectionItem {
     }
     return true;
   }
-}
+}
```
Comment (frederikspang) [inline lib/widgets/settings_widgets/settings_section_colorThemeButton.dart]: Fjernet newline?

File: lib/widgets/settings_widgets/settings_theme_display_box.dart [hunk 9]
```diff
@@ -43,4 +43,4 @@ class ThemeBox extends StatelessWidget {
       ),
     );
   }
-}
+}
```
Comment (frederikspang) [inline lib/widgets/settings_widgets/settings_theme_display_box.dart]: Fjernet icons?

File: pubspec.yaml [hunk 10]
```diff
@@ -25,7 +25,6 @@ dependencies:
   image: ^2.0.7
   quiver: '>=2.0.0 <3.0.0'
   mutex: ^1.0.1
-  material_design_icons_flutter: ^3.2.3895
```
Comment (frederikspang) [inline pubspec.yaml]: Bruger vi ikke disse ikoner længere?

File: test/screens/weekplan_selector_screen_test.dart [hunk 11]
```diff
@@ -166,162 +195,57 @@ void main() {
     expect(find.byKey(const Key('weekYear')), findsNWidgets(2));
   });
 
-  testWidgets('Should not have Redigér and Slet buttons outside edit mode',
```
Comment (frederikspang) [inline test/screens/weekplan_selector_screen_test.dart:167]: Er det meningen at I har slettet en del tests her?

**Review by frederikspang — CHANGES_REQUESTED** [review]

**Review by Lortemanden — CHANGES_REQUESTED** [review]
Now when you try to copy a weekplan to a citizen where there already is a weekplan with the same week and year there is no prompt to change the weekplan. Instead it just copies the weekplan onto the citizen. 

The "ændr" popup should still be present.

**Review by Lortemanden — DISMISSED** [review]

**Comment by nicolaibjerring** [comment]
We found a bug when testing the feature, where when you copy a week plan to another citizen it sometimes removes or adds activities from/to the copied week plan, e.g. it is not the same content that is copied. Attached is a recording of the bug. 
[copyBug.zip](https://github.com/aau-giraf/weekplanner/files/4597615/copyBug.zip)

**Review by nicolaibjerring — DISMISSED** [review]
We found a bug when testing the feature, where when you copy a week plan to another citizen it sometimes removes or adds activities from/to the copied week plan, e.g. it is not the same content that is copied. Attached is a recording of the bug.
[copyBug.zip](https://github.com/aau-giraf/weekplanner/files/4597929/copyBug.zip)

**Review by frederikspang — DISMISSED** [review]

File: test/screens/copy_to_citizens_screen_test.dart [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    await tester.pumpWidget(
+        MaterialApp(home: CopyToCitizensScreen(mockWeek, mockUser)));
+    await tester.pumpAndSettle();
+
+    expect(find.byKey(const Key('AcceptButton')), findsOneWidget);
+    expect(find.byKey(const Key('CancelButton')), findsOneWidget);
+  });
+
+  testWidgets(
+      'Test whether the it copies to citizesn when pressing the accept key',
```
Comment (Markusjeppesen) [inline test/screens/copy_to_citizens_screen_test.dart]:
```suggestion
      'Test whether it copies to citizens when pressing the accept button',
```

**Review by Markusjeppesen — CHANGES_REQUESTED** [review]

**Review by Markusjeppesen — APPROVED** [review]

**Review by Lortemanden — APPROVED** [review]

**Pull request merged**